### ISSUE TITLE
General questions about C coding

### ISSUE THREAD

**Issue by tamplan** [mate-desktop/mate-desktop#399]
Hello everyone,

Not really an issue but questions about Mate development in C language and i don't where to post this subject.

I have made some PR accepted by coping/pasting the same piece of code several times and sc0w have made this https://github.com/mate-desktop/mate-polkit/pull/49 whose code could be picked up elsewhere due to Gtk deprecated functions in others repositories whose use too GtkStock to replace.

Is it possible to imagine a new repository called, for example, **Mate Gtk** or something else to regroup several pieces of code about Gtk deprecated functions in a shared library called, for example, **libmategtk** used by various over Mate repositories?

What are the pro and con about this solution in term of size of applications on disk and memory used by the Mate Desktop project?

This should be a good place to regroup code and prepare future depreciation using GTK 4.0 for example.

Thanks by advance for any post about this subject.

Regards

**Comment by vkareh** [comment]
We regularly have conversations about this. One of the issues with this is that deprecated code is still shipped, which means any such effort cannot be used until MATE moves to GTK4, and at that time even the code that provided whatever functionality might no longer be compatible with the rest of GTK. Another issue is that of maintenance, this is a huge undertaking and who will take care of it?

The simple answer is yes, we will fork parts of GTK as needed to provide necessary functionality. We don't yet know when or whom or how, though.

**Comment by vkareh** [comment]
Case in point: mouse scrolling on notebook tabs: [GNOME/gtk@ad48f4d5](https://gitlab.gnome.org/GNOME/gtk/commit/ad48f4d52bbac6139dd829fcc421ad16441f34d2). this is a huge piece of functionality that was removed and there has been a lot of work done to bring it back to MATE.

It's not a simple "clone GtkNotebook into another repo" solution. It requires reimplementing this function in every program that needs it. Otherwise we would need to modify the widget class or provide an interface which will not be compatible with GtkNotebook moving forward, plus wouldn't work with any application using GtkNotebook anyway, so it still needs a per-application code change.

Ideally we can, at the libmate-desktop level, make _all_ GtkNotebook instances listen to the scroll event and call our own clone of `gtk_notebook_scroll`, but 1) how do you do that in a way that works for all GtkNotebooks in the exact same way, taking into account their widget container hierarchy? and 2) sounds like that would break a lot of applications.

Most deprecation/removals in GTK follow a similar pattern: functionality slowly gets dropped just because someone make a decision without thinking about consequences (read the bug that triggered removing the scrolling notebooks, it's actually quite painful to read: [630226](https://bugzilla.gnome.org/show_bug.cgi?id=630226)). Once enough functionality is dropped, the thing becomes harder to maintain than the functionality it provides, and so it gets marked as deprecated and removed in the next major release. There's no single commit for "remove this thing" that we can just fork and reuse, unfortunately. :(

**Comment by tamplan** [comment]
To be completely clear, it's a proposal.
Imagine this scenario:
Here https://github.com/mate-desktop/mate-polkit/pull/49/files we have a function named **polkit_mate_dialog_add_button** used to avoid GtkStock, a nice piece of code that can be copy/paste in other modules/applications of the Mate project.

Now we can imagine the same function but named **mate_gtk_dialog_add_button** (for example) and stored in a **libmategtk** shared library (the new repository), linked in **src/polkitmateauthenticationdialog.c**

I not speaking about rewrite any widget, just a set of usefull and  well documented functionsavailable and linked from a new dependency shared library.

We want to avoid GtkStock in mate-terminal, link it with **libmategtk** and use 
**mate_gtk_dialog_add_button**.

We want to avoid GtkStock in mate-calc, link it with **libmategtk** and use **mate_gtk_dialog_add_button**.

One place for code called from various applications to reduce size on disk and in memory usage i presume.

I hope i am more clean now for this proposal because English it's not my native language.

**Comment by vkareh** [comment]
Thanks for the clarification, this makes a lot of sense.

We have the mate-desktop repo for this purpose, and we can definitely add new useful functionality there to be reused throughout our projects :)

**Comment by tamplan** [comment]
Ok, thanks!
I study the case of the **src/polkitmateauthenticationdialog.c** for example...

**Comment by sc0w** [comment]
@tamplan  or maybe with the function to scroll tabs with mouse wheel, seems you use the same function in a lot of places in PRs :)

**Comment by tamplan** [comment]
Yes of course :)

**Comment by tamplan** [comment]
I have done this PR https://github.com/mate-desktop/mate-desktop/pull/400 to test this idea.

**Comment by raveit65** [comment]
Keep in mind that we have a lot of packages which doesn't have a direct dependency to mate-desktop.
EG. mate-polkit, mate-terminal, mate-system-monitor, pluma , and some more.
They can be used without installing mate-desktop under a different desktop for example.

So, do we really want to blow up dependencies only for some little code snipset?

I recall that it was nice and practicable to have packages without a dependency to mate-desktop, during port from gk2 to gtk3.
In result we could switch mate-terminal and others to gtk3 before having the whole desktop ready for gtk3 ;)

And gtk4 will come...., not yet , but maybe in some years.

**Comment by tamplan** [comment]
I did not think about using mate-terminal, for example, in another desktop environment than Mate Desktop.

But this is why i purpose initially the creation of a new repository named **mate-gtk** for a shared library **libmategtk** in charge of theses problems and only theses (avoid GtkStock, notebook scroll tab with mouse wheel and so on), every thing in relation with gtk used with Mate Desktop environment or not.

I am totally new in a C language contribution to Linux projects with Github, but not in programming (PHP, VBA, C 30 years ago :) ). Whenever I see a piece of code used several times in many places, I create a function to optimize and regroup the code in a project for example.

Maybe, here it's different with several projects, i presume too a dependency is not the better coding solution unless it's a dependency that we manage completely.

Hope this whole message is good translated from French to English :)

**Comment by vkareh** [comment]
@tamplan yes, maybe a new repo would be useful, but @raveit65 brings up a good point about dependencies. Other desktops (XFCE, and I think maybe Unity) use some of our software, so we shouldn't force mate-specific dependencies on them.

Is there any common dependency, like mate-common or something? I think whatever gtk forward-ports we do shouldn't be mate-specific, but gtk-specific. This way we keep it as simple as possible and it helps prevent pulling in other mate dependencies. Plus, other DEs could benefit from it too.

Regarding your English, don't worry too much about it. The majority of us are not native English speakers, we get along just fine ;)

**Comment by tamplan** [comment]
Afterwards, I am aware that the addition of a dependency implies the management of a new package by the different distributions, maybe, the simplest solution is to continue to copy/paste piece of code to solve problems about gtk through different applications of the Mate Desktop project.

**Comment by sc0w** [comment]
maybe in the future we can make new repo with the name `mtk` (mate tollkit) instead `gtk` (gimp toolkit) with forking all the needed and useful gtk deprecated code and adapted to work, and probably other desktops witch use `gtk` are interested in contribute over it too.

**Comment by yetist** [comment]
Create a new repo for common code is good idea, but I think use git submodule will be better for dependency.

**Comment by h1z1** [comment]
Why the push everywhere to GTK4 at all??  Those who use mate specifically do not care about any more then gnomeOfTheDay.
